exchanges/stream/websocket_test.go

```diff
@@ -0,0 +1,956 @@
+package stream
+
+import (
+	"bytes"
+	"compress/flate"
+	"compress/gzip"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/protocol"
+)
+
+const (
+	websocketTestURL = "wss://www.bitmex.com/realtime"
+	useProxyTests    = false                     // Disabled by default. Freely available proxy servers that work all the time are difficult to find
+	proxyURL         = "http://212.186.171.4:80" // Replace with a usable proxy server
+)
+
+var dialer websocket.Dialer
+
+type testStruct struct {
+	Error error
+	WC    WebsocketConnection
+}
+
+type testRequest struct {
+	Event        string          `json:"event"`
+	RequestID    int64           `json:"reqid,omitempty"`
+	Pairs        []string        `json:"pair"`
+	Subscription testRequestData `json:"subscription,omitempty"`
+}
+
+// testRequestData contains details on WS channel
+type testRequestData struct {
+	Name     string `json:"name,omitempty"`
+	Interval int64  `json:"interval,omitempty"`
+	Depth    int64  `json:"depth,omitempty"`
+}
+
+type testResponse struct {
+	RequestID int64 `json:"reqid,omitempty"`
+}
+
+var defaultSetup = &WebsocketSetup{
+	Enabled:                          true,
+	AuthenticatedWebsocketAPISupport: true,
+	WebsocketTimeout:                 time.Second * 5,
+	DefaultURL:                       "testDefaultURL",
+	ExchangeName:                     "exchangeName",
+	RunningURL:                       "wss://testRunningURL",
+	Connector:                        func() error { return nil },
+	Subscriber:                       func(_ []ChannelSubscription) error { return nil },
+	UnSubscriber:                     func(_ []ChannelSubscription) error { return nil },
+	GenerateSubscriptions: func() ([]ChannelSubscription, error) {
+		return []ChannelSubscription{
+			{Channel: "TestSub"},
+			{Channel: "TestSub2"},
+			{Channel: "TestSub3"},
+			{Channel: "TestSub4"},
+		}, nil
+	},
+	Features: &protocol.Features{Subscribe: true, Unsubscribe: true},
+}
+
+func TestTrafficMonitorTimeout(t *testing.T) {
+	ws := *New()
+	err := ws.Setup(defaultSetup)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ws.ShutdownC = make(chan struct{})
+	err = ws.trafficMonitor()
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Deploy traffic alert
+	ws.TrafficAlert <- struct{}{}
+
+	// Check if traffic monitor has started
+	err = ws.trafficMonitor()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Start new instance then simulate shutdown
+	err = ws.trafficMonitor()
+	if err != nil {
+		t.Fatal(err)
+	}
```

Comment:
This test doesn't do what you describe. Removing the wg.Wait highlighted this. You return nil if the trafficmonitor has already started, so there is no way to detect whether it has started. I think it is valid to return an error when that occurs. I also think you should simulate shutdown like you say, so you can prove that it exists